Lib/zipimport.py

```diff
@@ -347,114 +347,121 @@ def _read_directory(archive):
         raise ZipImportError(f"can't open Zip file: {archive!r}", path=archive)
 
     with fp:
+        # GH-87235: On macOS all file descriptors for /dev/fd/N share the same
+        # file offset, reset the file offset after scanning the zipfile diretory
+        # to not cause problems when some runs 'python3 /dev/fd/9 9<some_script'
+        start_offset = fp.tell()
         try:
-            fp.seek(-END_CENTRAL_DIR_SIZE, 2)
```

Comment:
The PR looks messy here. I just indented the body of "with fp" one level deeper to add the code to reset the file offset, that is:

```python3
start_offset = fp.tell()
try:
    <ORIGINAL CODE>
finally:
     fp.seek(start_offset)
```